domain/resource/service/resource.go

```diff
@@ -545,6 +566,101 @@ func (s *Service) SetRepositoryResources(
 	return s.st.SetRepositoryResources(ctx, args)
 }
 
+// AddResourcesBeforeApplication adds the details of which resource
+// revision to use before the application exists in the model. The
+// charm and resource metadata must exist. These resources are resolved
+// when the application is created using the returned Resource UUIDs.
+//
+// The following error types can be expected to be returned:
+//   - [resourceerrors.CharmIDNotValid] is returned if the Charm ID is not valid.
+//   - [resourceerrors.ArgumentNotValid] is returned if the origin is store and the revision
+//     is empty.
+//   - [resourceerrors.ResourceNameNotValid] is returned if resource name is empty.
+//   - [resourceerrors.ApplicationNameNotFound] if the specified application does
+//     not exist.
+func (s *Service) AddResourcesBeforeApplication(ctx context.Context, arg resource.AddResourcesBeforeApplicationArgs) ([]coreresource.UUID, error) {
+	if err := arg.CharmUUID.Validate(); err != nil {
+		return nil, errors.Errorf("%w: %w", resourceerrors.CharmIDNotValid, err)
+	}
+	if !isValidApplicationName(arg.ApplicationName) {
+		return nil, errors.Errorf("application name : %w", resourceerrors.ApplicationNameNotValid)
+	}
+	for _, res := range arg.ResourceDetails {
+		if res.Name == "" {
+			return nil, errors.Errorf("resource name is empty: %w", resourceerrors.ResourceNameNotValid)
+		}
+		if res.Origin == charmresource.OriginStore && res.Revision == nil {
```

Comment:
This could still do with an `res.Origin.Validate`.

Reply:
@hmlanigan this one still needs to be resolved, can you add it as a flyby in a subsequent PR?